File: 2015/python/day1-not-quite-lisp/santa.py
if __name__ == "__main__":
    ret, num = 0, 0
    basement = False

    with open('input.txt') as f:
        while True:
            char = f.read(1)          
            if not char:
                break
            if char == '(':
                ret += 1
            elif char == ')':
                ret -= 1
                if ret == -1 and not basement:
                    print (f'basement entered at position {num+1}')
                    basement = True
            else:
                pass
            num += 1
    print (f'final floor after {num} characters is {ret}')

File: 2015/cpp/day1/santa.cpp
#include <iostream>
#include <fstream>

int process(char ch) {
    if (ch == '(') { return 1; } else if (ch == ')') { return -1; } else return 0;
}

int main() {
    std::ifstream file("input.txt");

    if (!file.is_open()) {
        std::cerr << "Error: Could not open file." << std::endl;
        return 1;
    }
    int num = 0;
    int floor = 0;
    bool basement = false;

    char ch;
    while (file.get(ch)) {
        num += 1;
        floor += process(ch);
        if (!basement && floor < 0) {
            basement = true;
            std::cout << "basement entered at position " << num << std::endl;
        }
    }
    std::cout << "final floor after " << num << " characters is " << floor << std::endl;
    file.close();
    return 0;
}